### DISCUSSION TITLE
Easier way to wrap mdx components for use with theme UI

### DISCUSSION THREAD

**Discussion by genox** [tinacms/tinacms#3368] [Help]
Hi,

I am using theme-ui.com to style my nextjs page. In order to style content from tina, I need to wrap the text in theme UI components so they pick up formatting defined on theme level. I started extending mdx components as discussed for custom components in the docs and it works more or less but I'm reaching a point where I need a few suggestions:

1)  is there a simpler way to achieve what I need without having to write essentially a function for each html tag?
2)  is there a way to process components recursively that you suggest, e.g. to catch all the inline tags like `<a>` etc.?

It would be nice if I wouldn't have to care about processing things recursively, instead just defining "if you get a `<p>`, use `<Paragraph>`" no matter whether it contains an inline tag like `<a>` or `<strong>`. Just a simple map, essentially.

I assume that this affects every css-in-js approach on one or another level. Basically every approach that circumvents global css.

**Comment by jeffsee55** [comment]
> Just a simple map, essentially.

This is how it's intended to work, maybe I'm not understanding your use case properly. Can you provide some examples?

**Reply by genox** [reply]
I wonder what the most efficient approach is.

Right now I cover the basic html tags. Each block element that might contain an inline tag needs to be recursively rendered again to cover everything.

In essence, this is what it looks like. 

```
const ParagraphMdxInline: React.FC<{ nodes: any[] }> = memo(({ nodes }) => {
  return (
    <>
      {nodes.map((item: any) => {
        switch (item.type) {
          default:
            if (item.bold && item.italic) {
              return (
                <strong key={hash(item)}>
                  <em>{item.text}</em>
                </strong>
              );
            }
            if (item.bold) {
              return <strong key={hash(item)}>{item.text}</strong>;
            }
            if (item.italic) {
              return <em key={hash(item)}>{item.text}</em>;
            }
            return <React.Fragment key={hash(item)}>{item.text}</React.Fragment>;
        }
      })}
    </>
  );
});

const ParagraphMdx: React.FC<{ props: any }> = memo(({ props }) => {
  return (
      <Paragraph>
        {props.children.props.content.map((item: any) => {
          switch (item.type) {
            case 'a':
              return (
                <NextLink key={hash(item)} href={item.url} title={item.title} legacyBehavior>
                  <Link href={item.text}>
                    <ParagraphMdxInline nodes={item.children} />
                  </Link>
                </NextLink>
              );
            default:
              return <ParagraphMdxInline nodes={[item]} key={hash(item)} />;
          }
        })}
      </Paragraph>
  );
});

export const tinaMDXComponents: Components<any> = {
  p: (props) => {
    if (!props?.children.props.content[0].text) return <></>;
    return <ParagraphMdx props={props} />;
  },
};

```

Add to that h1..h5, ordered and unordered lists, etc and there is a lot of ground to cover. I could make the inline tag rendering function generic and call it every time.

I wonder if there is an approach that I don't see right now that allows simply defining any tag on `tinaMDXComponents` level without having to manage rendering children of block level items?

Or how would you approach this in an efficient manner? Just feels like a lot of rendering. Well I guess that's partially the tradeoff of the css-in-js buy in. ;)

With the way things are right now, I'm never really sure if I catch every possible occurrence of every possible tag.

**Reply by jeffsee55** [reply]
`props.children` shouldn't need to be touched directly, you should be able to just render it out. So `ParagraphMdx` should just be (in your example, `props.children` for the paragraph component will contain the `a` element you've overridden already):

```jsx
const ParagraphMdx: React.FC<{ props: any }> = memo(({ props }) => {
  return (
      <Paragraph>
        {props.children}
      </Paragraph>
  );
});

// and then override the `a` component, too:
export const tinaMDXComponents: Components<any> = {
  p: (props) => {
    return <ParagraphMdx props={props} />;
  },
  a: props => {
    return (
      <NextLink key={hash(item)} href={item.url} title={item.title} legacyBehavior>
       <Link href={item.text}>
         {props.children}
       </Link>
     </NextLink>
   );
};


```
This can be done for every component, here's a simple example of adding some classes to each element:

```jsx
        <TinaMarkdown
          content={props.content}
          components={{
            code_block: (props) => <Prism {...props} />,
            Hero: (props) => {
              return <Json src={props} />
            },
            bold: (props) => <strong className="text-red-300" {...props} />,
            italic: (props) => <em className="text-green-300" {...props} />,
            p: (props) => <p className="border" {...props} />,
            a: (props) => <a className="font-bold text-blue-600" {...props} />,
            // keep going until you've replaced all of them
          }}
        />
```
If you're using Typescript you should be able to click through to see all of the elements we use. For reference: https://github.com/tinacms/tinacms/blob/main/packages/tinacms/src/rich-text/index.tsx#L16-L43

**Reply by genox** [reply]
Ah, nice. Yeah I had a feeling I was going about this way too complicated. Thanks for your support!
